save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/ContestEnroller.kt

```diff
@@ -0,0 +1,197 @@
+
+    /**
+     * Name of the project that is planned to be enrolled for a contest
+     *
+     * Either organizationName and projectName or contestName should be set
+     */
+    var projectName: String?
+
+    /**
+     * Name of a contest that the chosen project should be enrolled for
+     *
+     * Either organizationName and projectName or contestName should be set
+     */
+    var contestName: String?
+
+    /**
+     * Lambda invoked when got response for enrollment request
+     */
+    var onResponseChanged: (String) -> Unit
+}
+
+/**
+ * Modal that shows [contestEnrollerComponent]
+ *
+ * @param isModalOpen flag that indicates whether the modal should be shown or not
+ * @param selectedContestName name of a current contest (should be null if [selectedProjectName] or [selectedOrganizationName] is not null)
+ * @param selectedOrganizationName name of an organization that project is from (should be null if [selectedContestName] is not null)
+ * @param selectedProjectName name of a current project (should be null if [selectedContestName] is not null)
+ * @param onResponse callback invoked when enrollment response is received
+ * @param onCloseButtonPressed callback invoked when close button was clicked
+ */
+@Suppress("TOO_MANY_PARAMETERS", "LongParameterList")
+fun ChildrenBuilder.showContestEnrollerModal(
+    isModalOpen: Boolean,
+    selectedContestName: String?,
+    selectedOrganizationName: String?,
+    selectedProjectName: String?,
+    onResponse: (String) -> Unit,
+    onCloseButtonPressed: () -> Unit,
+) {
+    modal { props ->
+        props.isOpen = isModalOpen
+        contestEnrollerComponent {
+            contestName = selectedContestName
+            projectName = selectedProjectName
+            organizationName = selectedOrganizationName
+            onResponseChanged = onResponse
+        }
+        div {
+            className = ClassName("d-sm-flex align-items-center justify-content-center mt-4")
+            button {
+                className = ClassName("btn btn-primary")
+                type = ButtonType.button
+                onClick = {
+                    onCloseButtonPressed()
+                }
+                +"Cancel"
+            }
+        }
+    }
+}
+
+/**
+ * A [FC] for a card that allows user to:
+ *  1. Choose a projects that should be enrolled for a contest
+ *  2. Choose a contest that the project should be enrolled for
+ *
+ * @return a functional component representing a contest enroller
+ */
+@Suppress(
+    "LongMethod",
+    "EMPTY_BLOCK_STRUCTURE_ERROR",
+    "TOO_LONG_FUNCTION",
+    "MAGIC_NUMBER",
+    "ComplexMethod",
+)
+private fun contestEnrollerComponent() = FC<ContestEnrollerProps> { props ->
+    val (organizationName, setOrganizationName) = useState(props.organizationName)
+    val (projectName, setProjectName) = useState(props.projectName)
+    val (contestName, setContestName) = useState(props.contestName)
+
+    val (availabilityUrl, _) = useState(contestName?.let {
```

Comment:
Does it need to be in state?

Reply:
Seems like yes because if it is not in state but is just a val, everything breaks after the first render